beacon_chain/networking/eth2_network.nim

```diff
@@ -71,19 +71,24 @@ type
     wantedPeers*: int
     hardMaxPeers*: int
     peerPool*: PeerPool[Peer, PeerId]
+    torpeerPool*:PeerPool[Peer, PeerId]
```

Comment:
Generic, as it were, comment relating to all of these `foo` and `torfoo` pairs: they create code duplication, which can be avoided by having an entire Tor switch, so it's be `torSwitch.peerPool`, etc.

As usual, it depends what the goals here are, how they're balanced between research and deployability.

But more or less all the remaining hundreds of lines of code changes here are duplications of functions resulting from this design decision.

```nim
proc torcheckPeer(node: Eth2Node, peerAddr: PeerAddr): bool =
  logScope: peer = peerAddr.peerId
  let peerId = peerAddr.peerId
  if node.torpeerPool.hasPeer(peerId):
    trace "Already connected over tor"
    false
  else:
    if node.istorSeen(peerId):
      trace "Recently connected over tor"
      false
    else:
      # check if the normal switch is already using the same connection
      # avoid any overlap with normal peer use to prevent leakage or correlation
      if node.peerPool.hasPeer(peerId) : 
        echo "Avoiding Overlap of tor with normal's ", peerAddr
        false
      else:
        true
```
isn't meaningfully different from `checkPeer` except that it refers to `torpeerPool`, rather than `peerPool`, and calls `istorSeen` rather than `isSeen.`

There's some notion of that it has to know about the `peerPool` too, but in general, there are more precise ways to satisfy most of these constraints.

So this is the question: as research code, sure, why not.

I'd push against incorporating this design into actual-`nimbus-eth2`, if that were proposed.